### DISCUSSION TITLE
FastAPI Not Returning 200 OK Response

### DISCUSSION THREAD

**Discussion by 4pygmalion** [fastapi/fastapi#13087] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
@thyroid_router.post("/predict", response_model=MILResponseModel)
async def predict(
    query: RequestModel,
    mil_predictor: MILPredictor = Depends(get_mil_predictor),
    config=Depends(get_config),
    logger: Logger = Depends(get_logger),
) -> MILResponseModel:


    logger.info(
        f"Accept request slide_path({query.slide_path}), Thread ID({threading.get_ident()})"
    )

    deepzoom_level = get_deepzoom_level(config["mpp"], OpenSlide(query.slide_path))
    tiler = Tiler(
        config.tile_size,
        config.overlap,
        config.limit_bounds,
        deepzoom_level=deepzoom_level,
        n_workers=config.tile_workers,
        patch_filter=PatchFilter(
            [
                check_small_compressed_format,
            ]
        ),
        logger=logger,
    )
    output_queue: queue.Queue = tiler.do_tile_otsu(query.slide_path)
    tiler.join()
    patches = Patches.from_queue(output_queue)
    logger.info(f"End of tiling N={len(patches)} ")

    logger.debug("Start of slide_level prediction")
    confidences, attention_weights = mil_predictor.predict(patches)
    logger.debug("End of prediction")

    # slide-evel prediction
    logger.debug("Confidence to float")
    confidences = confidences.ravel()
    max_idx = confidences.ravel().argmax()
    slide_level_label: str = BinaryLabels(max_idx).name
    slide_confidence = float(confidences[max_idx])

    logger.debug("Build response model")
    slide = OpenSlide(query.slide_path)
    deepzoom_level: int = get_deepzoom_level(config.mpp, slide)
    slide_result = MILSlideResult(
        confidence=slide_confidence,
        label=slide_level_label,
    )
    response = MILResponseModel.build(
        query=query,
        slide_result=slide_result,
        patches=patches,
        heatmap_scores=attention_weights,
        patch_mpp=config.mpp,
        model_uuid=config.model_uuid,
        model_name=mil_predictor.mil_model.__class__.__name__,
        top_patch_percent=config.top_patch_percent,
    )
    logger.info(
        f"End of request slide_path({query.slide_path}), Thread ID({threading.get_ident()})"
    )

    return response
```

#### Summary
A FastAPI application running in a Docker container does not return a 200 OK response, even though the process completes without any exceptions. The container logs show no errors, and the API appears to be running, but no successful responses are observed.

#### Symptoms
- No 200 OK response for valid requests.
- API requests sent to the correct port (e.g. 30052) and endpoint.
- The process inside the container completes successfully.
- No exceptions or errors appear in the container logs.

### Operating System

Linux

### FastAPI Version

0.111.0

### Pydantic Version

2.7.1

### Python Version

Python 3.8.19

### Additional Context

![image](https://github.com/user-attachments/assets/730cc6e4-6950-4d50-9565-426a5ce54b25)

Note that "121 INFO, end of tiling N=..." refer to the completion of process above screenshot

**Comment by luzzodev** [comment, marked as answer]
Hello @4pygmalion there are some inconsistencies between your code and the attached screen.
What error do you have on client? Are you running the docker container behind a proxy?
Based on the screen, the response is not sent starting at 02:07:22 is it right?
Additional: You are running a blocking not async code in the main loop

**Reply by 4pygmalion** [reply]
Thank you for reply @luzzodev,

Here are my answers for your questions,


> 1. there are some inconsistencies between your code and the attached screen.

I have updated my screenshot to ensure consistency between my code and the screenshot.
![image](https://github.com/user-attachments/assets/b6e3a6b6-3402-4ef0-b7bd-3d6fd02dd541)

> 2. The client received any response because the request was timeouted.

The client used the following code. It waits for API responses for up to 10 minutes, but if the time limit is exceeded, it enters the catch block at:
`var response = await httpClient.PostAsync(API, data);`
and returns the error message: "A task was cancelled."

There is no additional error message indicating the server status.
![image](https://github.com/user-attachments/assets/279f23ea-51b6-4255-a472-85c15d92edf6)

> 3. Are you running the docker container behind a proxy?

No, I am not using any proxy server between the Docker container and the client.

> 4. Based on the screen, the response is not sent starting at 02:07:22 is it right?

Yes, that is correct. The response was not sent starting at that time.

> Additional: You are running a blocking not async code in the main loop

Yes, I am aware of that. This is because the line
`output_queue: queue.Queue = tiler.do_tile_otsu(query.slide_path)`
requires significant computational resources, utilizing up to 48 CPUs.

I intentionally blocked the use of `async` to run this on a single thread because enabling this multiprocessing in multithread by FastAPI may cause a deadlock or timeout

**Reply by 4pygmalion** [reply]
Thank you @luzzodev, I found the client's request as as async. This problem is due to timeout on client side"Thank you, @luzzodev. I noticed that the client's request is asynchronous. The issue appears to be caused by a timeout on the client side."
